pkg/querier/queryrange/limits.go

```diff
@@ -453,6 +454,29 @@ func NewLimitedRoundTripper(next queryrangebase.Handler, limits Limits, configs
 	return transport
 }
 
+type SemaphoreWithTiming struct {
+	sem *semaphore.Weighted
+}
+
+func NewSemaphoreWithTiming(max int64) *SemaphoreWithTiming {
+	return &SemaphoreWithTiming{
+		sem: semaphore.NewWeighted(int64(max)), // Replace with your semaphore creation logic
+	}
+}
+
+// acquires the semaphore and records the time it takes.
+func (s *SemaphoreWithTiming) Acquire(ctx context.Context, n int64) (int64, error) {
+	start := time.Now()
+
+	if err := s.sem.Acquire(ctx, int64(n)); err != nil {
+		return 0, err
+	}
+
+	elapsed := time.Since(start)
+
+	return elapsed.Milliseconds(), nil
```

Comment:
I would just return `time.Since(start)` as the value here, and call `String()` on it when you log it so it'll show a useful value at any timescale.